dagshub/upload/wrapper.py

```diff
@@ -549,6 +561,66 @@ def get_repo_url(self, url_format, directory, branch=None):
             ),
         )
 
+    def upload_file_to_s3(self, s3_client, local_path, bucket_name, remote_path):
+        """
+        Upload a single file to S3.
+
+        :param s3_client: An instance of boto3 S3 client
+        :param local_path: The local path to the file
+        :param bucket_name: The S3 bucket name
+        :param remote_path: The S3 path where the file will be uploaded
+        """
+        s3_client.upload_file(local_path, bucket_name, remote_path)
+        log_message(f'Uploaded {os.path.relpath(local_path, os.getcwd())} to "{self._api.full_name}"...', logger)
+
+    def upload_files_to_bucket(
+        self,
+        local_path,
+        remote_path,
+        recursive=False,
+        max_workers=config.upload_threads,
+        **kwargs
+    ):
+        """
+        Upload a file or directory to an S3 bucket, preserving the directory structure.
+
+        :param local_path: Path to the local directory to upload
+        :param remote_path: The directory path within the S3 bucket
+        :param recursive: True if this is a directory upload, or False if single file
+        :param max_workers: The maximum number of threads to use
+        """
+        file_exceptions = ['.DS_Store']  # Default list of files to skip
+
+        s3 = get_repo_bucket_client(self._api.full_name)
+        if remote_path.split("/")[0] == self.name:
+            remote_path = remote_path.split("/", 1)[1]
+        if not recursive:
+            self.upload_file_to_s3(s3_client=s3, local_path=local_path, bucket_name=self.name, remote_path=remote_path)
```

Comment:
This isn't what `recursive` usually means.

The logic should be:
```python
if is_file(local_path):
  upload_file(file)
if is_dir(local_path):
  if recursive:
    <the os.walk you have written>
  else:
    for file in os.listdir(local_path):
      upload_file(file)
```

I recommend looking at the `download_files` function for a way to make it more manageable, basically splitting the process into two steps: 1) get the list of things to upload and where to upload them, 2) actually upload them